core/src/main/java/com/graphhopper/routing/profiles/IntEncodedValue.java

```diff
@@ -0,0 +1,148 @@
+ * creates a new entry. Either the engine knows the higher level logic or we assume the default value is 0 and
+ * map this value to the real value on every retrieval request.
+ * <p>
+ * How could you then implement e.g. a 'priority' value going from [-3, 3] that maps to [0,7] but should
+ * have a default value of 3 instead of 0? Either you waste space and map this to [1,7], which means that 0 and 3 both
+ * refer to the same 0 value (currently the preferred method due to its simplicity) or you could create a
+ * MappedIntEncodedValue class that holds an array or a Map with the raw integers similarly to what StringEncodedValue does:
+ * {0: 0, 1: -3, 2: -2, 3: -1, 4: 1, 5: 2, 6: 3}
+ */
+public class IntEncodedValue implements EncodedValue {
+
+    private final String name;
+    // we do not have just one or two int values like with flags, we can have much more:
+    private int dataIndex;
+
+    final int bits;
+    int maxValue;
+    int fwdShift;
+    int bwdShift;
+    int fwdMask;
+    int bwdMask;
+    int defaultValue;
+    boolean store2DirectedValues;
+
+    public IntEncodedValue(String name, int bits) {
+        this(name, bits, 0, false);
+    }
+
+    /**
+     * @param defaultValue defines which value to return if the 'raw' integer value is 0.
+     */
+    public IntEncodedValue(String name, int bits, int defaultValue, boolean store2DirectedValues) {
+        this.name = name;
+        if (!name.toLowerCase().equals(name))
+            throw new IllegalArgumentException("EncodedValue name must be lower case but was " + name);
+
+        this.bits = bits;
+        if (bits <= 0)
+            throw new IllegalArgumentException("bits cannot be 0 or negative");
+        this.defaultValue = defaultValue;
+        this.store2DirectedValues = store2DirectedValues;
+    }
+
+    @Override
+    public final void init(EncodedValue.InitializerConfig init) {
+        this.dataIndex = init.dataIndex;
+
+        this.fwdShift = init.shift;
+        this.fwdMask = init.next(bits);
+        if (store2DirectedValues) {
+            this.bwdShift = init.shift;
+            this.bwdMask = init.next(bits);
+        }
+
+        this.maxValue = (1 << bits) - 1;
+    }
+
+    private void checkValue(int value) {
+        if (fwdMask == 0)
+            throw new IllegalStateException("EncodedValue " + getName() + " not initialized");
+        if (value > maxValue)
+            throw new IllegalArgumentException(name + " value too large for encoding: " + value + ", maxValue:" + maxValue);
+        if (value < 0)
+            throw new IllegalArgumentException("negative value for " + name + " not allowed! " + value);
+    }
+
+    /**
+     * This method 'merges' the specified integer value with the specified 'flags' to return a value that can
+     * be stored.
+     *
+     * @return the storable format that can be read via fromStorageFormatToInt
+     */
+    public final int toStorageFormat(boolean reverse, int flags, int value) {
```

Comment:
Currently it is an array of ints as the underlying RAMDataAccess is an array of ints and so this is the most efficient. But I'm thinking of using a byte array instead, which would be required if the GH node or way IDs exceed the 2^32 integer boundaries. The best would be if we could encapsulate this a bit with something like the IntRefs class and pass this in this method, which would also avoid the need of the public getOffset method.